core/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryUnitTests.java

```diff
@@ -127,4 +146,78 @@ public void testFilterNonMasterPingResponse() {
             assertThat(filteredNodes, equalTo(allNodes));
         }
     }
+
+    public void testNodesUpdatedAfterClusterStatePublished() throws Exception {
+        ThreadPool threadPool = new TestThreadPool(getClass().getName());
+        Settings settings = Settings.builder()
+                                // randomly make minimum_master_nodes a value higher than we have nodes for, so it will force failure
+                                .put(DISCOVERY_ZEN_MINIMUM_MASTER_NODES_SETTING.getKey(), randomBoolean() ? "3" : "1").build();
+
+        Map<String, MockNode> nodes = new HashMap<>();
+        ZenDiscovery zenDiscovery = null;
+        ClusterService clusterService = null;
+        try {
+            Set<DiscoveryNode> expectedFDNodes = null;
+            // create master node and its mocked up services
+            MockNode master = createMockNode("master", settings, null, threadPool, logger, nodes).setAsMaster();
+            ClusterState state = master.clusterState; // initial cluster state
+
+            // build the zen discovery and cluster service
+            clusterService = createClusterService(threadPool, master.discoveryNode);
+            setState(clusterService, state);
+            zenDiscovery = buildZenDiscovery(settings, master, clusterService, threadPool);
+
+            // a new cluster state with a new discovery node (we will test if the cluster state
+            // was updated by the presence of this node in NodesFaultDetection)
+            MockNode newNode = createMockNode("new_node", settings, null, threadPool, logger, nodes);
+            ClusterState newState = ClusterState.builder(state).incrementVersion().nodes(
+                DiscoveryNodes.builder(state.nodes()).add(newNode.discoveryNode).masterNodeId(master.discoveryNode.getId())
+            ).build();
+
+            try {
+                // publishing a new cluster state
+                ClusterChangedEvent clusterChangedEvent = new ClusterChangedEvent("testing", newState, state);
+                AssertingAckListener listener = new AssertingAckListener(newState.nodes().getSize() - 1);
+                expectedFDNodes = zenDiscovery.getFaultDetectionNodes();
+                zenDiscovery.publish(clusterChangedEvent, listener);
+                listener.await(1, TimeUnit.HOURS);
+                // publish was a success, update expected FD nodes based on new cluster state
+                expectedFDNodes = fdNodesForState(newState, master.discoveryNode);
+            } catch (Discovery.FailedToCommitClusterStateException e) {
+                // not successful, so expectedFDNodes above should remain what it was originally assigned
```

Comment:
can we capture whether we set min master node to 3 or 1 and check here that the exception is really expected?